test/metabase/api/dataset_test.clj

```diff
@@ -107,6 +108,44 @@
                   :running_time true}
                  (format-response (most-recent-query-execution-for-query query)))))))))
 
+(deftest override-metadata-query-test
+  (testing "POST /api/dataset"
+    (testing "when `:dataset-metadata` is passed in the body, make sure the returned columns metadata get combined with it (#22517)"
+      (let [dataset-metadata [{:name          "ID"
+                               :display_name  "VENUES ID CAPITALIZED"
+                               :description   "normal venues id but capitalized"
+                               :semantic_type "type/FK" ;; originally is type/PK
+                               :base_type     "type/BigInteger"
+                               :field_ref     ["field" (mt/id :venues :id) nil]}]
+            query            (assoc (mt/mbql-query venues {:fields [$id]})
+                                    :dataset-metadata dataset-metadata)]
+        (is (= dataset-metadata
+               (->> (mt/user-http-request :rasta :post 202 "dataset" query)
+                    mt/cols
+                    (map #(select-keys % [:name :display_name :description :base_type :field_ref :semantic_type])))))
+
+        (testing "We shouldn't keep the dataset-metadata inside query when running it"
+          (let [run-query-async api.dataset/run-query-async]
+            (with-redefs [api.dataset/run-query-async (fn [query & opts]
+                                                        (assert (nil? (:dataset-metadata query)))
+                                                        (run-query-async query opts))]
+              (mt/user-http-request :rasta :post 202 "dataset" query))))))
+
+    (testing "should run successfully even if dataset-metadata contains fields that is not in the query"
+      (let [;; We're querying venues but this dataset-metadata contains a user-id
+            ;; This is to simulate the cases where `dataset-metadata` sent from FE
+            ;; contains fields that are not resolved anywhere in our query.
+            ;; Since we're allowing this behavior of sending custom dataset-metadata
+            ;; we expect the query to run successfully.
+            dataset-metadata [{:name          "ID"
+                               :display_name  "USER ID CAPITALIZED"
+                               :description   "User id"
+                               :semantic_type "type/FK"
+                               :base_type     "type/BigInteger"
+                               :field_ref     ["field" (mt/id :users :id) nil]}]]
+        (mt/user-http-request :rasta :post 202 "dataset" (assoc (mt/mbql-query venues {:fields [$id]})
+                                                          :dataset-metadata dataset-metadata))))))
```

Comment:
Shouldn't there be some sort of `is` assertion to check the results? Doesn't seem like this isn't really testing anything